tensorflow_io/avro/20190307-avro-dataset.md

```diff
@@ -0,0 +1,299 @@
+
+We can design this thread safe w/out locks by creating a new parser instance for each step of work; if we like to share context information that is valid across multiple calls, we will pass that as argument.
+
+For efficiency and feature completeness this parser leverages an ordered prefix tree where the prefixes are defined through common ancestors in the avro data structure. This prefix tree must be ordered to support filter expressions. The parts of a filter expression have to be evaluated before being able to evaluate filter itself.
+
+#### Value buffer (single instance per thread)
+
+Assembles avro parsed data into a TensorFlow tensor.
+
+Handles filling in of missing values using default scalar and tensor data.
+
+Handles adding of a dimension for batching.
+
+Not used concurrently, e.g. no reading/writing occurs concurrently. Each thread holds its own instance exclusively. We can divide work across multiple tensors or slices of a tensor. Slices can be generated in a map and then an assembler thread can pick up these slices to create the entire tensor.
+
+#### Avro dataset
+
+Maps tensors to TensorFlow&#39;s dataset interface.
+
+Provides the bridge to the python code.
+
+### Code Design Python
+
+Supports Python 2.x and 3.x.
+
+The code is organized into two parts: 1) the Avro dataset and 2) a make method that creates the dataset. For testing we leverage the avro python library and the avro python-3 library with some thin wrappers to support Python 2.x and 3.x transparently.  Warning: Python 2.x and Python 3.x represents strings differently. Here, we treat Python 3.x as first-class citizen and provide backward compatible support to python 2.x whenever necessary.
+
+#### Avro dataset
+
+The bridge to the native implementation following examples of other datasets such as the Example dataset that leverages protobuf or the csv dataset.
+
+#### Make Avro dataset
+
+A helper method that fills in defaults for most parameters of the dataset. In addition, the make method takes care of batching, shuffling, and prefetching.
+
+## Alternative designs
+- Rather than providing a langauge for mapping all avro into TensorFlow; once could have defined a fixed schema of TensorFlow types in avro. An example for this is the Example format in TensorFlow that leverages protobuf. The above design provides more flexibility and was chosen for that reason.
+- Could we have mapped any avro type into TensorFlow?  This would have been limited to mapping arrays to dense tensors and some form of fixed record structures into sparse tensors. The implicit assumptions of what is supported vs. not supported and how the mapping is done would have been too confusing.
+
+
+## Questions and Discussion Topics
+- Should we include building avroc into scope?  Additional effort is transforming a cmake script into bazel build config. This would enable running tests and guards against code quality digression.
```

Comment:
I would prefer to include avroc. This could avoid may issues compared to linking .so. I took a brief look at avroc. Looks like it depends on zlib, lzma, snappy & jansson. I have some experience before with most of them. Think it should be possible to include avroc. I can help with bazel file if needed.

Reply:
Agreed. Let's aim for that.